server/middleware/cached_test.go

```diff
@@ -0,0 +1,67 @@
+package middleware
+
+import (
+	"time"
+
+	"github.com/cloudwan/gohan/schema"
+	"github.com/golang/mock/gomock"
+	"github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/rackspace/gophercloud"
+)
+
+var _ = ginkgo.Describe("Cached identity service", func() {
+
+	var (
+		cachedIdentityService IdentityService
+		mockedIdentityService *MockIdentityService
+		auth                  schema.Authorization
+		serviceClient         *gophercloud.ServiceClient
+		tenantID              string
+		tenantName            string
+		token                 string
+	)
+
+	ginkgo.BeforeEach(func() {
+		t := ginkgo.GinkgoT()
+		mockedIdentityService = NewMockIdentityService(gomock.NewController(t))
+		cachedIdentityService = NewCachedIdentityService(mockedIdentityService, time.Second)
+		serviceClient = &gophercloud.ServiceClient{ProviderClient: &gophercloud.ProviderClient{TokenID: token}}
+		tenantID = "tenant-id"
+		tenantName = "tenant-name"
+		token = "token"
+		auth = schema.NewAuthorization(tenantID, tenantName, token, []string{}, []*schema.Catalog{})
+	})
+
+	ginkgo.It("Use inner service if authorization is not cached and save returned one in cache", func() {
+		mockedIdentityService.EXPECT().VerifyToken(token).Return(auth, nil).Times(1)
+		rv, err := cachedIdentityService.VerifyToken(token)
+		Expect(rv).To(Equal(auth))
+		Expect(err).To(BeNil())
+		rv, err = cachedIdentityService.VerifyToken(token)
+		Expect(rv).To(Equal(auth))
+		Expect(err).To(BeNil())
+	})
+
+	ginkgo.It("Pass GetTenantID to inner service", func() {
+		mockedIdentityService.EXPECT().GetTenantID(tenantName).Return(tenantID, nil)
+		rv, err := cachedIdentityService.GetTenantID(tenantName)
+		Expect(rv).To(Equal(tenantID))
+		Expect(err).To(BeNil())
+	})
+
+	ginkgo.It("Pass GetTenantName to inner service", func() {
+		mockedIdentityService.EXPECT().GetTenantName(tenantID).Return(tenantName, nil)
+		rv, err := cachedIdentityService.GetTenantName(tenantID)
```

Comment:
the inner service is actually a `keystoneV3Client`, it will send at least one HTTP request per each call. have you measured what's the performance overhead and if that's acceptable?
same issue applies to GetTenantId

Reply:
Added caching for GetTenantName since it's used in each request. GetTenantID is not used anywhere.